### ISSUE TITLE
Do() should panic when given function that returns values

### ISSUE THREAD

**Issue by ash2k** [uber-go/mock#75]
I've just spent an hour debugging a test where I by mistake used `Do(func () string { ... return "some value" })` instead of `DoAndReturn()`.

`Do()` should error out ASAP when given a function that returns values.

Thank you for the library!

**Comment by r-hang** [comment]
If I understand your issue correctly, this proposal turns a current no-op operation into a panic, which may break lots of existing tests that rely on this no-op behavior. One way to address this seems to be some type of linting to prevent this.

**Comment by ash2k** [comment]
Yes. There may also be tests that pass but don't actually test what the author thinks they are testing.

**Comment by Tulzke** [comment]
Maybe it's worth excluding the return values from the function passed to Do? It's about the case when we pass the `-typed` flag to mockgen.
Example:
```go
type SomeInterface{
    Action(value int) (bool, error)
}

func (MockSomeInterface) Do(func(value int)) {...}

func (MockSomeInterface) DoAndReturn(func(value int) (bool, error)) {...}
```
